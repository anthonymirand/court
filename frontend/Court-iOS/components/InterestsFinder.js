import React from 'react';
import PropTypes from 'prop-types';
import { Icon } from 'expo';

import { Platform, StyleSheet, Text, TextInput, TouchableOpacity, View } from 'react-native';
import Colors from '../constants/Colors';

import InterestsCloud from '../components/InterestsCloud';

import { queryGoogleKG } from '../utils/GoogleKGS';

/**
* Provides interface for searching for and selecting interests
*/
export default class InterestsFinder extends React.Component {
  constructor(props) {
    super(props);
    this.state = { query: '', results: [] };
  }

  searchCount = 1;

  handleQuery(query) {
    const countBefore = ++this.searchCount;
    // Query Google KnowledgeGraph
    let results = [];
    if (query !== '') {
      queryGoogleKG(query).then((result) => {
        // Another search happened before the response was done; cancel.
        if (countBefore !== this.searchCount) {
          return
        }
        for (var i = 0; i < result.length && i < 10; i++) {
          const title = result[i].result.name;
          const description = result[i].result.description;
          const id = result[i].result['@id'];
          results.push({title: title, description: description, id: id, imgUrl: result[i].result.image ? result[i].result.image.contentUrl : null });
        }
        this.setState({ query, results });
        return;
      });
    }
    this.setState({ query, results });
  }

  render() {
    const { query, results } = this.state;
    const { interests, onAddInterest, onRemoveInterest } = this.props;

    return (
      // Select your gender
      <View style={styles.outerView}>
        <View style={styles.textInputWrapper} >
          <TextInput
            multiline={false}
            placeholder="The Office"
            placeholderColor="rgba(255,255,255,0.5)"
            style={styles.textInput}
            onChangeText={(query) => this.handleQuery(query)}
            value={this.state.query}
          />
          <Icon.Ionicons
            name='ios-search'
            size={35}
            color={Colors.peach}
          />
        </View>
        {
          <View style={styles.results}>
            <InterestsCloud interests={interests} recommendations={results} onAddInterest={onAddInterest} onRemoveInterest={onRemoveInterest}/>
          </View>
        }
      </View>
    );
  }
}

InterestsFinder.propTypes = {
  /**
  * Object containing the user's current selected interests
  */
  interests: PropTypes.object.isRequired,
  /**
  * Callback function for when a user selects a new interest
  */
  onAddInterest: PropTypes.func.isRequired,
  /**
  * Callback function for when a user removes an interest
  */
  onRemoveInterest: PropTypes.func.isRequired,
}

const styles = StyleSheet.create({
  outerView: {
    marginVertical: 15,
    marginHorizontal: 20,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
  },
  textInputWrapper: {
    ...Platform.select({
      ios: {
        shadowColor: 'black',
        shadowOffset: { height: 8 },
        shadowOpacity: 0.15,
        shadowRadius: 9,
      },
      android: {
        elevation: 20,
      },
    }),
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'white',
    borderWidth: 1,
    borderColor: 'white',
    borderRadius: 15,
    paddingVertical: 10,
    paddingHorizontal: 15,
  },
  textInput: {
    fontSize: 30,
    paddingTop: 5,
    fontFamily: 'orkney-medium',
    color: Colors.peach,
    flex: 1,
  },
  results: {
    flex: 1,
  },
});
