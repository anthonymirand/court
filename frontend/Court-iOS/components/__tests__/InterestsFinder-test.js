import 'react-native';
import React from 'react';
import InterestsFinder from '../InterestsFinder';
import renderer from 'react-test-renderer';

it('Renders Default InterestsFinder', () => {
  const mocks = { onAddInterest: jest.fn(), onRemoveInterest: jest.fn() };
  const interests = {'kgs://12hv3': {title: 'The Office', description: 'American Sitcom', imgUrl: 'test.url'}};

  const tree = renderer.create(
    <InterestsFinder
      interests={interests}
      onAddInterest={mocks.onAddInterest}
      onRemoveInterest={mocks.onRemoveInterest}
    />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
