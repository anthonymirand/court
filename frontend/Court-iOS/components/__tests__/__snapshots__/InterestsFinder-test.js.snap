// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Default InterestsFinder 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "column",
      "justifyContent": "center",
      "marginHorizontal": 20,
      "marginVertical": 30,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomWidth": 1,
        "borderColor": "white",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      multiline={false}
      onChangeText={[Function]}
      placeholder="The Office"
      placeholderColor="rgba(255,255,255,0.)"
      style={
        Object {
          "color": "white",
          "flex": 1,
          "fontSize": 30,
        }
      }
      value=""
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    />
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "center",
          "paddingTop": 20,
        }
      }
    />
  </View>
</View>
`;
